File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid/Presenter/SharedView/SharedView_PopupControlNumber.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Icu.Text;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.App;
using Android.Text;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Droid.Class;
using BPMOPMobile.Droid.Core.Common;
using BPMOPMobile.Droid.Presenter.Fragment;
using Newtonsoft.Json.Linq;

namespace BPMOPMobile.Droid.Presenter.SharedView
{
    class SharedView_PopupControlNumber : SharedView_PopupControlBase
    {
        public SharedView_PopupControlNumber(LayoutInflater _inflater, AppCompatActivity _mainAct, CustomBaseFragment _fragment, string _fragmentTag, View _rootView) : base(_inflater, _mainAct, _fragment, _fragmentTag, _rootView)
        {

        }

        public override void InitializeValue_Master(ViewElement clickedElement)
        {
            base.InitializeValue_Master(clickedElement);
        }

        public override void InitializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JObject JObjectChild)
        {
            base.InitializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
        }

        public override void InitializeView()
        {
            base.InitializeView();
            try
            {
                #region Get View - Init Data
                View _viewPopupControl = _inflater.Inflate(Resource.Layout.PopupControl_Number, null);
                ImageView _imgBack = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_Number_Close);
                TextView _tvTitle = _viewPopupControl.FindViewById<TextView>(Resource.Id.tv_PopupControl_Number_Title);
                ImageView _imgDone = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_Number_Done);
                ImageView _imgDelete = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_Number_Delete);
                EditText _edtContent = _viewPopupControl.FindViewById<EditText>(Resource.Id.edt_PopupControl_Number);
                ImageView _imgClearText = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_Number_ClearText);

                //_edtContent.InputType = InputTypes.ClassNumber; // Only Number
                _imgDelete.Visibility = ViewStates.Gone;
                _edtContent.InputType = InputTypes.ClassNumber | InputTypes.NumberFlagDecimal; // Only Number
                _edtContent.SetFilters(new Android.Text.IInputFilter[] { new Android.Text.InputFilterLengthFilter(255) }); // Limit lại 255 kí tự

                switch (base.flagView)
                {
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                    default:
                        {
                            _tvTitle.Text = elementParent.Title;

                            CultureInfo _culVN = CultureInfo.GetCultureInfo("vi-VN");
                            CultureInfo _culEN = CultureInfo.GetCultureInfo("en-US");

                            if (!String.IsNullOrEmpty(elementParent.Value))
                            {
                                double _customValue = double.Parse(elementParent.Value, _culEN);
                                DecimalFormat df = new DecimalFormat("0");
                                df.MaximumFractionDigits = int.MaxValue;
                                _edtContent.Text = df.Format(_customValue);
                            }
                            else
                                _edtContent.Text = "";

                            break;
                        }
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                        {
                            _tvTitle.Text = elementPopup.Title;

                            CultureInfo _culVN = CultureInfo.GetCultureInfo("vi-VN");
                            CultureInfo _culEN = CultureInfo.GetCultureInfo("en-US");

                            if (!String.IsNullOrEmpty(elementPopup.Value))
                            {
                                double _customValue = double.Parse(elementPopup.Value, _culEN);
                                DecimalFormat df = new DecimalFormat("0");
                                df.MaximumFractionDigits = int.MaxValue;
                                _edtContent.Text = df.Format(_customValue);
                            }
                            else
                                _edtContent.Text = "";

                            break;
                        }
                }
                #endregion

                #region Show View                
                Dialog _dialogPopupControl = new Dialog(_rootView.Context, Resource.Style.Theme_Custom_BPMOP_Dialog_FullScreen);
                Window window = _dialogPopupControl.Window;
                _dialogPopupControl.RequestWindowFeature(1);
                _dialogPopupControl.SetCanceledOnTouchOutside(false);
                _dialogPopupControl.SetCancelable(true);
                window.SetSoftInputMode(SoftInput.AdjustPan); // Để Hide soft keyboard ko bị Resize
                window.SetGravity(GravityFlags.Bottom);
                var dm = _mainAct.Resources.DisplayMetrics;

                _dialogPopupControl.SetContentView(_viewPopupControl);
                _dialogPopupControl.Show();
                WindowManagerLayoutParams s = window.Attributes;
                s.Width = WindowManagerLayoutParams.MatchParent;
                s.Height = WindowManagerLayoutParams.MatchParent;
                window.Attributes = s;
                #endregion

                #region Event
                _edtContent.TextChanged += (sender, e) =>
                {

                };
                _edtContent.AfterTextChanged += (sender, e) =>
                {
                    //double _customValue;
                    //CultureInfo _culVN = CultureInfo.GetCultureInfo("vi-VN");
                    //string _currentTextValue = _edtContent.Text.Trim()/*.Replace(".", "")*/;

                    //if (_currentTextValue.Contains("E+"))
                    //{
                    //    if (double.TryParse(_currentTextValue, out _customValue))
                    //    {
                    //        if (!_customValue.ToString("N0", _culVN).Equals(_edtContent.Text))
                    //        {
                    //            _edtContent.Text = _customValue.ToString("N0", _culVN);
                    //        }
                    //    }
                    //}
                    //else
                    //{
                    //    if (double.TryParse(_currentTextValue.Replace(".", ""), out _customValue))
                    //    {
                    //        if (!_customValue.ToString("N0", _culVN).Equals(_edtContent.Text))
                    //        {
                    //            _edtContent.Text = _customValue.ToString("N0", _culVN);
                    //        }
                    //    }
                    //}
                    //_edtContent.SetSelection(_edtContent.Text.Length); // focus vào character cuối cùng
                };
                _imgBack.Click += (sender, e) =>
                {
                    CmmDroidFunction.HideSoftKeyBoard(_edtContent, _mainAct);
                    _dialogPopupControl.Dismiss();
                };
                _imgClearText.Click += (sender, e) =>
                {
                    _edtContent.Text = "";
                };

                _imgDelete.Click += (sender, e) =>
                {
                    _edtContent.Text = "";
                    _imgDone.PerformClick();
                };

                _imgDone.Click += delegate
                {
                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                            {
                                CmmDroidFunction.HideSoftKeyBoard(_edtContent, _mainAct);

                                string _result = "";
                                if (!String.IsNullOrEmpty(_edtContent.Text))
                                {
                                    CultureInfo _culEN = CultureInfo.GetCultureInfo("en-US");
                                    double _customValue = double.Parse(_edtContent.Text.Trim().Replace(',','.'), _culEN);

                                    if (!String.IsNullOrEmpty(elementParent.DataSource)) // _element.DataSource trả ra là số chữ số decimal
                                    {
                                        int _demical = int.Parse(elementParent.DataSource);
                                        _result = Math.Round(_customValue, _demical > 0 ? _demical : 0).ToString();
                                    }
                                    else
                                        _result = Math.Round(_customValue, 0).ToString();
                                }

                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, _result);
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                CmmDroidFunction.HideSoftKeyBoard(_edtContent, _mainAct);

                                string _result = "";
                                if (!String.IsNullOrEmpty(_edtContent.Text))
                                {
                                    CultureInfo _culEN = CultureInfo.GetCultureInfo("en-US");
                                    double _customValue = double.Parse(_edtContent.Text.Trim().Replace(',', '.'), _culEN);

                                    if (!String.IsNullOrEmpty(elementPopup.DataSource)) // _element.DataSource trả ra là số chữ số decimal
                                    {
                                        int _demical = int.Parse(elementPopup.DataSource);
                                        _result = Math.Round(_customValue, _demical > 0 ? _demical : 0).ToString();
                                    }
                                    else
                                        _result = Math.Round(_customValue, 0).ToString();
                                }

                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, _result);
                                _dialogPopupControl.Dismiss();
                                break;
                            }
                    }
                };
                #endregion

                _edtContent.Text = _edtContent.Text;
                _edtContent.RequestFocus();
                CmmDroidFunction.ShowSoftKeyBoard(_rootView, _mainAct);
            }
            catch (Exception ex)
            {
#if DEBUG
                CmmDroidFunction.WriteTrackingError(this.GetType().Name, "InitializeView", ex);
#endif
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/shareview/SharedView_PopupControlNumber.java
package com.vuthao.bpmop.shareview;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.text.Editable;
import android.text.InputFilter;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.gson.Gson;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.keyboard.KeyboardManager;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.core.Vars;
import com.vuthao.bpmop.detail.custom.DetailFunc;

import org.json.JSONObject;

import java.text.DecimalFormat;

public class SharedView_PopupControlNumber extends SharedView_PopupControlBase implements TextWatcher, View.OnClickListener {
    private Dialog dialog;
    private View view;
    private ImageView imgBack;
    private TextView tvTitle;
    private ImageView imgDone;
    private ImageView imgDelete;
    private EditText edtContent;
    private ImageView imgClearText;

    public SharedView_PopupControlNumber(LayoutInflater inflater, Activity mainAct, String fragmentTag, View rootView) {
        super(inflater, mainAct, fragmentTag, rootView);
    }

    @Override
    public void initializeValue_Master(ViewElement viewElement) {
        super.initializeValue_Master(viewElement);
    }

    @Override
    public void initializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JSONObject JObjectChild) {
        super.initializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
    }

    @Override
    public void initializeView() {
        super.initializeView();

        view = inflater.inflate(R.layout.control_popup_number, null);
        imgBack = view.findViewById(R.id.img_PopupControl_Number_Close);
        tvTitle = view.findViewById(R.id.tv_PopupControl_Number_Title);
        imgDone = view.findViewById(R.id.img_PopupControl_Number_Done);
        imgDelete = view.findViewById(R.id.img_PopupControl_Number_Delete);
        edtContent = view.findViewById(R.id.edt_PopupControl_Number);
        imgClearText = view.findViewById(R.id.img_PopupControl_Number_ClearText);

        imgDelete.setVisibility(View.GONE);
        edtContent.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_DECIMAL);
        edtContent.setFilters(new InputFilter[]{new InputFilter.LengthFilter(255)});

        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                tvTitle.setText(elementParent.getTitle());

                if (!Functions.isNullOrEmpty(elementParent.getValue())) {
                    long l = Long.parseLong(elementParent.getValue());
                    DecimalFormat df = new DecimalFormat("0");
                    df.setMaximumFractionDigits(Integer.MAX_VALUE);
                    edtContent.setText(df.format(l));
                } else {
                    edtContent.setText("");
                }

                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                tvTitle.setText(elementPopup.getTitle());

                if (!Functions.isNullOrEmpty(elementPopup.getValue())) {
                    long l = Long.parseLong(elementPopup.getValue());
                    DecimalFormat df = new DecimalFormat("0");
                    df.setMaximumFractionDigits(Integer.MAX_VALUE);
                    edtContent.setText(df.format(l));
                } else {
                    edtContent.setText("");
                }
                break;
            }
        }

        // region Dialog
        dialog = new Dialog(mainAct, R.style.Theme_Custom_BPMOP_Dialog_FullScreen);
        Window window = dialog.getWindow();
        dialog.requestWindowFeature(1);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setCancelable(true);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        window.setGravity(Gravity.BOTTOM);
        dialog.setContentView(view);
        dialog.show();

        WindowManager.LayoutParams s = window.getAttributes();
        s.width = WindowManager.LayoutParams.MATCH_PARENT;
        s.height = WindowManager.LayoutParams.MATCH_PARENT;
        window.setAttributes(s);
        // endregion

        imgBack.setOnClickListener(this);
        edtContent.addTextChangedListener(this);
        imgClearText.setOnClickListener(this);
        imgDelete.setOnClickListener(this);
        imgDone.setOnClickListener(this);

        edtContent.setText(edtContent.getText().toString());
        edtContent.requestFocus();
        KeyboardManager.showKeyBoard(edtContent, mainAct);
    }

    private void done() {
        KeyboardManager.hideKeyboard(edtContent, mainAct);
        String result = "";

        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                if (!Functions.isNullOrEmpty(edtContent.getText().toString())) {
                    double d = Double.parseDouble(edtContent.getText().toString().trim());
                    result = DetailFunc.share.getFormatControlDecimal(d, elementParent);
                }

                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATEFORM);
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", result.replaceAll("\\,", ""));
                BroadcastUtility.send(mainAct, intent);
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                if (!Functions.isNullOrEmpty(edtContent.getText().toString())) {
                    double d = Double.parseDouble(edtContent.getText().toString().trim());
                    result = DetailFunc.share.getFormatControlDecimal(d, elementPopup);
                }

                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATECHILDACTION);
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", result.trim().replaceAll("\\,", ""));
                BroadcastUtility.send(mainAct, intent);
                break;
            }
        }

        dialog.dismiss();
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void afterTextChanged(Editable editable) {

    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.img_PopupControl_Number_Close: {
                KeyboardManager.hideKeyboard(edtContent, mainAct);
                dialog.dismiss();
                break;
            }
            case R.id.img_PopupControl_Number_ClearText: {
                edtContent.setText("");
                break;
            }
            case R.id.img_PopupControl_Number_Delete: {
                edtContent.setText("");
                imgDone.performClick();
                break;
            }
            case R.id.img_PopupControl_Number_Done: {
                done();
                break;
            }
        }
    }
}
